# Анализ базы данных мобильного приложения

## Описание проекта

Компания преобрела крупный сервис для чтения книг по подписке и теперь нуждается в анализе данных для дальнейшего построения бизнеса. 

**Основная цель исследования** - провести анализ базы данных, содержащей информацию о книгах, издательствах, авторах,  пользовательских обзорах и оценках, чтобы в дальнейшем сформулировать ценностное предложение для нового продукта, который предоставляет сервис по чтению книг по подписке.


###  Загрузка библиотек и изучение общей информации

Чтобы приступить к анализу, необходимо загрузить необходимые для дальнейшей работы библиотеки.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

Настроим доступ к базе данных, установив необходимые параметры для этого.

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Теперь можем изучить информацию о данных, для этого выведим таблицы на экран.

Для качественной предобработки данных создадим универсальную функцию information(), которая будет на вход принимать датафрейм, а на выходе будет выводить все необходимые характеристики, поочередно применяя методы head, info, duplicated и isna.

In [3]:
#пропишем функцию
def information(df):
    display(df.head())
    df.info()
    display(df.isna().sum())
    display(df.duplicated().value_counts())

Выведим первые 5 строк таблицы **'books'**'

In [4]:
query = '''SELECT * 
FROM books 
'''
con=engine.connect()
books = pd.io.sql.read_sql(sql=text(query), con = con)
information(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64

False    1000
dtype: int64

Выведим первые 5 строк таблицы **''authors'**'

In [5]:
query = '''SELECT * 
FROM authors 
'''
con=engine.connect()
authors = pd.io.sql.read_sql(sql=text(query), con = con)
information(authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


author_id    0
author       0
dtype: int64

False    636
dtype: int64

Выведим первые 5 строк таблицы **''publishers'**' 

In [6]:
query = '''SELECT * 
FROM publishers 
'''
con=engine.connect()
publishers = pd.io.sql.read_sql(sql=text(query), con = con)
information(publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


publisher_id    0
publisher       0
dtype: int64

False    340
dtype: int64

Выведим первые 5 строк таблицы **''ratings'**' 

In [7]:
query = '''SELECT * 
FROM ratings 
'''
con=engine.connect()
ratings = pd.io.sql.read_sql(sql=text(query), con = con)
information(ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


rating_id    0
book_id      0
username     0
rating       0
dtype: int64

False    6456
dtype: int64

Выведим первые 5 строк таблицы **''reviews'**' 

In [8]:
query = '''SELECT * 
FROM reviews 
'''
con=engine.connect()
reviews = pd.io.sql.read_sql(sql=text(query), con = con)
information(reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


review_id    0
book_id      0
username     0
text         0
dtype: int64

False    2793
dtype: int64

Итак, все данные загружены, дубликатов и пропусков в данных нет, данные в столбцах отображены корректно. 

##  Задание 1

Необходимо посчитать, сколько книг вышло после 1 января 2000 года.

In [9]:
query = '''SELECT COUNT(book_id) 
FROM books 
WHERE publication_date::date > '01-01-2000'
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0    819

**819** книг вышло после 1 января 2000 года

##  Задание 2

Для каждой книги необходимо посчитать количество обзоров и среднюю оценку.

In [10]:
query = '''SELECT b.book_id,
                  b.title,
                  ROUND(AVG(ra.rating ), 2) AS avg_rating,
                  COUNT(DISTINCT r.review_id) AS count_review
FROM books AS b
LEFT JOIN reviews AS r ON b.book_id = r.book_id
LEFT JOIN ratings AS ra ON r.book_id = ra.book_id
GROUP BY b.book_id
ORDER BY count_review DESC
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                              title  avg_rating  \
0        948                            Twilight (Twilight  #1)        3.66   
1        963                                Water for Elephants        3.98   
2        734                                   The Glass Castle        4.21   
3        302  Harry Potter and the Prisoner of Azkaban (Harr...        4.41   
4        695  The Curious Incident of the Dog in the Night-Time        4.08   
..       ...                                                ...         ...   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel         NaN   
996      808                            The Natural Way to Draw         NaN   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites         NaN   
998      221                          Essential Tales and Poems         NaN   
999      191  Disney's Beauty and the Beast (A Little Golden...         NaN   

     count_review  
0               7  
1               6  
2               6  
3               6  
4               6  
..            ...  
995             0  
996             0  
997             0  
998             0  
999             0  

[1000 rows x 4 columns]

Итак, получилась таблица с средней оценкой каждой книги, отсортированная по убыванию количества обзоров.

Можно заметить, что книга Twilight (Twilight #1) получила наибольшее число обзоров. 

##  Задание 3

Необходимо определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры.

In [11]:
query = '''SELECT p.publisher,
           COUNT(DISTINCT b.book_id) AS count_books          
FROM books AS b
JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE num_pages > 50
GROUP BY p.publisher
ORDER BY count_books DESC
LIMIT 1
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  count_books
0  Penguin Books           42

Издательство 'Penguin Books' выпустило больше всего книг - 42 книги. 

##  Задание 4

Необходимо определить автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [12]:
query = '''SELECT author,
           ROUND(AVG(rating), 2) AS avg_rating
FROM books AS b
JOIN ratings AS r ON b.book_id = r.book_id 
JOIN authors AS a ON b.author_id = a.author_id
WHERE b.book_id IN (SELECT book_id
                    FROM ratings
                    GROUP BY book_id
                    HAVING COUNT(*) >= 50)
GROUP BY author
ORDER BY avg_rating DESC
LIMIT 1
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

Автор J.K. Rowling/Mary GrandPré имеет самый высокий средний рейтинг книг среди книг с 50 и более оценок.

##  Задание 5

Необходимо посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [13]:
query = '''WITH data_1 AS (SELECT username,
                  COUNT(review_id) AS count_rew
FROM reviews
WHERE username IN (SELECT username
FROM ratings
GROUP BY username
HAVING COUNT(rating_id) > 48)
GROUP BY username)

SELECT ROUND(AVG(count_rew), 2) AS avg_count_rew
FROM data_1
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg_count_rew
0           24.0

Среднее количество обзоров от пользователей, которые поставили больше 48 оценок равно 24.

# Вывод

На основе проведенного анализа можно сделать следующие **выводы:**
- приложение имеет 819 книг, выпущенных после 1 января 2000 года
- посчитаны и собраны данные в единую таблицу о каждой книге, о количестве обзоров на произведение и высчитана средняя оценка
- издательство Penguin Books выпустило больше всего книг - 42
- автор J.K. Rowling/Mary GrandPré имеет самый высокий средний рейтинг книг 
- среднее количество обзоров от пользователей, которые поставили больше 48 оценок равно 24.

**Рекомендации:**
- продолжить активное сотрудничество с издательством Penguin Books, так как они выпустили наибольшее количество книг
- начать сотрудничать с авторами, имеющими высокие оценки за книги, чтобы продвигать их книги через рекламные кампании, тем самым можно удержать внимание пользователей в приложении и привлечь новых юзеров
- настроить рекомендательную систему для пользователей, чтобы для каждого пользователя был подбор книг по его интересам
- рассмотреть вариант с добавлением в приложение аудиокниг, чтобы привлечь новую категорию пользователей, которые любят слушать книги